tests/mock_crust/churn.rs

```diff
@@ -18,49 +18,83 @@
 use itertools::Itertools;
 use rand::Rng;
 use routing::{Authority, DataIdentifier, Event, EventStream, MessageId, QUORUM, Request, XorName};
-use routing::mock_crust::{Config, Network};
+use routing::mock_crust::{Config, Endpoint, Network};
 use std::cmp;
 use std::collections::{HashMap, HashSet};
-use super::{TestClient, TestNode, create_connected_clients, create_connected_nodes,
-            gen_range_except, poll_and_resend, verify_invariant_for_all_nodes};
+use super::{Nodes, TestClient, TestNode, create_connected_clients, create_connected_nodes,
+            gen_range_except, poll_all, poll_and_resend, verify_invariant_for_all_nodes};
 
-// Randomly add or remove some nodes, causing churn.
-// If a new node was added, returns the index of this node. Otherwise
-// returns `None` (it never adds more than one node).
+// Randomly remove some nodes.
 //
-// Note: it's necessary to call `poll_all` afterwards, as this function doesn't
-// call it itself.
-fn random_churn<R: Rng>(rng: &mut R,
-                        network: &Network,
-                        nodes: &mut Vec<TestNode>)
-                        -> Option<usize> {
+// Note: it's necessary to call `poll_all` afterwards, as this function doesn't call it itself.
+fn drop_random_nodes<R: Rng>(rng: &mut R, nodes: &mut Vec<TestNode>, min_section_size: usize) {
     let len = nodes.len();
+    // Nodes needed for quorum with minimum section size. Round up.
+    let min_quorum = (min_section_size * QUORUM + 99) / 100;
 
-    if len > network.min_section_size() + 2 && rng.gen_weighted_bool(3) {
-        let _ = nodes.remove(rng.gen_range(1, len));
-        let _ = nodes.remove(rng.gen_range(1, len - 1));
-        let _ = nodes.remove(rng.gen_range(1, len - 2));
+    if rng.gen_weighted_bool(3) {
+        // Pick a section then remove as many nodes as possible from it without breaking quorum.
+        let i = rng.gen_range(0, len);
+        let prefix = *nodes[i].routing_table().our_prefix();
 
-        None
-    } else {
-        let proxy = rng.gen_range(0, len);
-        let index = rng.gen_range(1, len + 1);
-        let config = Config::with_contacts(&[nodes[proxy].handle.endpoint()]);
+        // Any network must allow at least one node to be lost:
+        let num_excess = cmp::max(1,
+                                  cmp::min(nodes[i].routing_table().our_section().len() -
+                                           min_quorum,
+                                           len - min_section_size));
+        assert!(num_excess > 0);
 
-        nodes.insert(index, TestNode::builder(network).config(config).create());
-        Some(index)
+        let mut removed = 0;
+        // Remove as many more as possible:
```

Comment:
This comment still implies that at least one has _already_ been removed.  I guess I'd just say "Remove nodes" or something like that.